aiida_cp2k/workchains/multistage.py

```diff
@@ -0,0 +1,486 @@
+            nstages += 1
+    output_dict['nstages'] = nstages
+    output_dict['last_stage_tag'] = 'stage_{}'.format(nstages - 1)
+    output_dict['nruns_stage0'] = len(kwargs)
+    output_dict['stage_opt_converged'] = []
+    output_dict['stage_nsteps'] = []
+    output_dict['step_info'] = {}
+    step_info_list = [
+        'step', 'energy_au', 'dispersion_energy_au', 'pressure_bar', 'cell_vol_angs3', 'cell_a_angs', 'cell_b_angs',
+        'cell_c_angs', 'cell_alp_deg', 'cell_bet_deg', 'cell_gam_deg', 'max_step_au', 'rms_step_au', 'max_grad_au',
+        'rms_grad_au', 'scf_converged'
+    ]
+    for step_info in step_info_list:
+        output_dict['step_info'][step_info] = []
+    for istage in six.moves.range(nstages):  # avoid the stage0 with failing settings
+        key = stages_arg['stage_{}'.format(istage)]
+        kwarg = kwargs[key].get_dict()
+        output_dict['stage_opt_converged'].append(kwarg['motion_opt_converged'])
+        nsteps = kwarg['motion_step_info']['step'][-1]
+        output_dict['stage_nsteps'].append(nsteps)
+        for istep, step in enumerate(kwarg['motion_step_info']['step']):
+            # Exclude redoundant zeroth calculations but remember that LBFGS starts from 1!
+            if not (istage > 0 and step == 0):
+                for step_info in step_info_list:
+                    output_dict['step_info'][step_info].append(kwarg['motion_step_info'][step_info][istep])
+
+    # Outputs from the last stage only
+    output_dict['dft_type'] = kwargs[key].get_dict()['dft_type']
+    output_dict['bandgap_spin1_au'] = kwargs[key].get_dict()['bandgap_spin1_au']
+    output_dict['bandgap_spin2_au'] = kwargs[key].get_dict()['bandgap_spin2_au']
+
+    return Dict(dict=output_dict).store()
+
+
+class Cp2kMultistageWorkChain(WorkChain):
+    """Workchain to submit GEO_OPT/CELL_OPT/MD workchains with iterative fashion
+
+    NOTE:
+    1) There are different CP2K's input parameters that are updated in parallel:
+    - AttributeDict(self.exposed_inputs(Cp2kBaseWorkChain, 'base')) is the initial one
+      that contains the custom setting from the run.
+    - self.ctx.base_inp is a Dict that is copied from the previous one and updated.
+      It also contains metadata (e.g., labels) and can contain extra inputs for the base
+    - self.ctx.parameters is a dict that contains only cp2k parameters.
+      Just before the calculation it is merged into the previous one.
+    - self.ctx.parameters_yaml is a dict that contains the parameter from the protocol
+      and is used to update self.ctx.parameters et every change of settings/stage
+
+    """
+
+    _calculation_class = Cp2kCalculation
+
+    @classmethod
+    def define(cls, spec):
+        super(Cp2kMultistageWorkChain, cls).define(spec)
+
+        # Inputs
+        spec.expose_inputs(Cp2kBaseWorkChain,
+                           namespace='cp2k_base',
+                           exclude=['parameters', 'cp2k.metadata.options.parser_name'])
```

Comment:
```suggestion
                           exclude=['cp2k.parameters', 'cp2k.metadata.options.parser_name'])
```
I think it should be 'cp2k.parameters', right?

Reply:
Not sure, we need to look carefully together with the first comment on how to make the input parameter optional.

Reply:
I think if you exclude `cp2k.parameters`, then the input parameter (from the first comment) can remain `required=True`.

Reply:
If I didn't overlook anything, in the multistage work chain you anyways generate this input parameter on the fly

Reply:
In the test https://github.com/danieleongari/aiida-cp2k/blob/aiida1_cleanmerge/examples/workchains/test_multistage_h2o_fail.py
I do modify the parameters, and I like to keep the input like this so that an advanced user can always modify the final CP2K setting independently  from the protocol.

Reply:
The best here would be to create an input that is to use your suggestion and then have a `base_cp2k.cp2k.parameters` that is not required, so that we can keep `required=True` in the Cp2kCalc. However, I don't know if this is possible, i.e., to make a "nested" input.

Reply:
I saw this is possible:
```
        spec.expose_inputs(Cp2kBaseWorkChain,
                           namespace='cp2k_base',
                           exclude=['cp2k.parameters', 'cp2k.metadata.options.parser_name'])

        spec.input('cp2k_base.cp2k.parameters',
                   valid_type=Dict,
                   required=False,
                   help='Specify custom CP2K settings that will be overwritten just before submitting the calculation',
                   )
```